yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -142,6 +145,117 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
     val containerIdString = System.getenv(ApplicationConstants.Environment.CONTAINER_ID.name())
     ConverterUtils.toContainerId(containerIdString)
   }
+
+  /**
+   * Obtains token for the Hive metastore, using the current user as the principal.
+   * Some exceptions are caught and downgraded to a log message.
+   * @param conf hadoop configuration; the Hive configuration will be based on this
+   * @return a token, or `None` if there's no need for a token (no metastore URI or principal
+   *         in the config), or if a binding exception was caught and downgraded.
+   */
+  def obtainTokenForHiveMetastore(conf: Configuration): Option[Token[DelegationTokenIdentifier]] = {
+    try {
+      obtainTokenForHiveMetastoreInner(conf, UserGroupInformation.getCurrentUser().getUserName)
+    } catch {
+      case e: Exception => {
+        handleTokenIntrospectionFailure("Hive", e)
+        None
+      }
+    }
+  }
+
+  /**
+   * Handle failures to obtain a token through introspection. Failures to load the class are
+   * not treated as errors: anything else is.
+   * @param service service name for error messages
+   * @param thrown exception caught
+   * @throws Exception if the `thrown` exception isn't one that is to be ignored
+   */
+  private[yarn] def handleTokenIntrospectionFailure(service: String, thrown: Throwable): Unit = {
+    thrown match {
+      case e: ClassNotFoundException =>
+        logInfo(s"$service class not found $e")
+        logDebug("Hive Class not found", e)
+      case e: NoClassDefFoundError =>
+        logDebug(s"$service class not found", e)
+      case e: InvocationTargetException =>
+        // problem talking to the metastore or other hive-side exception
+        logInfo(s"$service method invocation failed", e)
+        throw if (e.getCause != null) e.getCause else e
+      case e: ReflectiveOperationException =>
+        // any other reflection failure log at error and rethrow
+        logError(s"$service Class operation failed", e)
+        throw e;
```

Comment:
`NoClassDefFound` may need to be covered too; I think it's the transient version of CNFE, though that may be superstition.

I'd like to still unwind the `InvocationTargetException`, as its just a wrapper for what came in before; cutting it will simply reduce one level of needless stack trace